drivers/i2c/i2c_dw.c

```diff
@@ -730,15 +847,26 @@ static int i2c_dw_transfer(const struct device *dev, struct i2c_msg *msgs, uint8
 			break;
 		}
 
+		if (dw->state & I2C_DW_ERR_MASK) {
+			if (dw->state & (I2C_DW_SDA_STUCK | I2C_DW_SCL_STUCK)) {
+				ret = -ETIME;
+			} else if (dw->state & (I2C_DW_NACK | I2C_DW_CMD_ERROR)) {
+				ret = -EIO;
+			}
+			k_sem_reset(&dw->device_sync_sem);
```

Comment:
Other I2C transfer errors are not resetting the semaphore.  I don't think this is needed on this path.